src/ess/loki/io.py

```diff
@@ -157,80 +161,51 @@ def merge_background_monitor_runs(
-    if sample_name is None:
-        sample_position = sc.vector(value=[0, 0, 0], unit='m')
-    else:
-        sample_position = snx.compute_positions(instrument[sample_name][()])['position']
-    return source_position, sample_position
-
-
-def _load_events(
-    filename: FilePath[Filename[RunType]],
-    data_name: Union[NeXusDetectorName, NeXusMonitorName[MonitorType]],
-    transform_path: TransformationPath,
-    source_name: NeXusSourceName,
-    sample_name: Optional[NeXusSampleName],
-) -> sc.DataGroup:
-    with snx.File(filename) as f:
-        instrument = f['entry'][NEXUS_INSTRUMENT_PATH]
-        dg = instrument[data_name][()]
-        source_position, sample_position = _load_source_and_sample_positions(
-            instrument, source_name, sample_name
-        )
-    dg = snx.compute_positions(dg, store_transform=transform_path)
+def load_nexus_sample(file_path: FilePath[Filename[RunType]]) -> RawSample[RunType]:
+    return RawSample[RunType](nexus.load_sample(file_path))
 
-    events = _preprocess_data(
-        dg[f'{data_name}_events'],
-        sample_position=sample_position,
-        source_position=source_position,
+
+def dummy_load_sample(file_path: FilePath[Filename[RunType]]) -> RawSample[RunType]:
+    return RawSample[RunType](
+        sc.DataGroup({'position': sc.vector(value=[0, 0, 0], unit='m')})
     )
-    if data_name in DETECTOR_BANK_RESHAPING:
-        events = DETECTOR_BANK_RESHAPING[data_name](events)
 
-    dg[f'{data_name}_events'] = events
-    return dg
+
+def load_nexus_source(file_path: FilePath[Filename[RunType]]) -> RawSource[RunType]:
+    return RawSource[RunType](nexus.load_source(file_path))
 
 
 def load_nexus_detector(
-    filename: FilePath[Filename[RunType]],
+    file_path: FilePath[Filename[RunType]],
     detector_name: NeXusDetectorName,
-    transform_path: TransformationPath,
-    source_name: NeXusSourceName,
-    sample_name: Optional[NeXusSampleName],
+    raw_source: RawSource[RunType],
+    raw_sample: RawSample[RunType],
 ) -> LoadedSingleFileDetector[RunType]:
-    return LoadedSingleFileDetector[RunType](
-        _load_events(
-            filename=filename,
-            data_name=detector_name,
-            transform_path=transform_path,
-            source_name=source_name,
-            sample_name=sample_name,
-        )
+    out = nexus.load_detector(file_path=file_path, detector_name=detector_name)
+    # Note here we specify the name instead of using
+    # ess.reduce.nexus.extract_detector_data because we need the name to put the
+    # events back into the original data group.
+    key = f'{detector_name}_events'
+    events = _preprocess_data(
+        out[key],
+        sample_position=raw_sample['position'],
+        source_position=raw_source['position'],
     )
+    if detector_name in DETECTOR_BANK_RESHAPING:
+        events = DETECTOR_BANK_RESHAPING[detector_name](events)
+    out[key] = events
```

Comment:
Can you explain your comment, and why we need this? Can't these transformations be applied after extracting the detector data?

Reply:
So you're saying we could insert an additional step after the `RawData[Sample]` has been extracted, where we would reshape and patch (add necessary coordinates, etc.) the data, and output something called `PreparedData` that would be just before the `MaskedData` in the chain?

Reply:
Yes.